File: insights/insights/page/suppliers_estimate/suppliers_estimate.js

frappe.pages['suppliers-estimate'].on_page_load = function(wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Suppliers for Estimation Needs',
        single_column: true
    });

    page.add_inner_button(__('Print Table'), function() {
        let table = document.querySelector('.supplier-insights-table');
        if (!table) {
            frappe.msgprint("No data to print");
            return;
        }
        let clone = table.cloneNode(true);

        // Remove all elements with class "details-btn"
        clone.querySelectorAll('.details-btn').forEach(btn => btn.remove());
    
        // Also remove the header cell if it exists
        clone.querySelectorAll('th, td').forEach(cell => {
            if (cell.textContent.trim() === "Details" || cell.querySelector('.details-btn')) {
                cell.remove();
            }
        });

        let print_window = window.open('', '', 'height=700,width=900');

        print_window.document.write(`
            <html>
            <head>
                <title>Suppliers for Estimation Needs</title>
                <style>
                    table {
                        width: 100%;
                        border-collapse: collapse;
                        font-size: 12px;
                    }
                    table, th, td {
                        border: 1px solid #000;
                    }
                    th, td {
                        padding: 6px;
                        text-align: left;
                    }
                    th {
                        background: #f2f2f2;
                    }
                </style>
            </head>
            <body>
                ${clone.outerHTML}
            </body>
            </html>
        `);

        print_window.document.close();
        print_window.focus();
        print_window.print();
    });

    page.add_inner_button(__('Export to Excel'), function() {
        let table = document.querySelector('.supplier-insights-table table');
        if (!table) {
            frappe.msgprint("No data to export");
            return;
        }
    
        let cloned = table.cloneNode(true);
    
        let detailsColIndex = -1;
        let headerCells = cloned.querySelectorAll("thead th, tr:first-child th");
        headerCells.forEach((th, i) => {
            if (th.innerText.trim().toLowerCase() === "details") {
                detailsColIndex = i;
            }
        });
    
        if (detailsColIndex > -1) {
            cloned.querySelectorAll("tr").forEach(tr => {
                let cells = tr.querySelectorAll("th, td");
                if (cells[detailsColIndex]) {
                    cells[detailsColIndex].remove();
                }
            });
        }
    
        cloned.querySelectorAll(".details-btn").forEach(btn => btn.remove());
    
        let csv = [];
        let rows = cloned.querySelectorAll("tr");
    
        rows.forEach(tr => {
            let row = [];
            let cols = tr.querySelectorAll("th, td");
            cols.forEach(td => {
                let htmlContent = td.innerHTML.replace(/<br\s*\/?>/gi, "\n");
                let text = (new DOMParser().parseFromString(htmlContent, "text/html")).body.textContent;
                text = text
                    .split("\n")
                    .map(line => line.trim()) 
                    .filter(line => line.length > 0)
                    .join("\n");
    
                row.push('"' + text.replace(/"/g, '""') + '"');
            });
            csv.push(row.join(","));
        });
    
        let blob = new Blob(["\ufeff" + csv.join("\n")], { type: "text/csv;charset=utf-8;" });
        let link = document.createElement("a");
        link.href = URL.createObjectURL(blob);
        link.download = "Suppliers_Estimate.csv";
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
    });    
    
    page.set_primary_action('Generate Report', function() {
        fetch_supplier_estimate(filters);
    });
    $(`<style>
        .dt-cell__content {
            text-align: left !important;
        }</style>
    `).appendTo(page.main);	

	let supplierParam = '';

	if (frappe.route_options && frappe.route_options.supplier) {
		supplierParam = frappe.route_options.supplier[0];
		frappe.route_options = null;
	} else {
		const urlParams = new URLSearchParams(window.location.search);
		supplierParam = urlParams.get('supplier');
	}

    let filters = {
        "supplier":[],
        "has_po_only":0,
        "has_sq_only":0
    };

    this.page = page;
    $(page.main).on("click", ".details-btn", function (e) {
        e.preventDefault();
        e.stopPropagation();
        console.log("Details button clicked!", this);
        let supplier_code = $(this).data("supplier");
        let doctype_name = $(this).data("doctype");
        let details = $(this).data("details");
        let supplier_name = $(this).data("supplier_name");
        let items = $(this).data("items");
        open_purchase_details(supplier_code, supplier_name, doctype_name, details, items);
    });
    this.form = new frappe.ui.FieldGroup({
        fields: [
            {
                fieldtype: 'MultiSelectList',
                label: 'Supplier',
                fieldname: 'supplier',
                options: "Supplier",
                get_data: function (txt) {
                    return frappe.db.get_link_options("Supplier", txt);
                },
                onchange: function() {
                    filters.supplier = this.values || [];
                }
            },
            {
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'Link',
                label: 'Supplier Group',
                fieldname: 'supplier_group',
                options: 'Supplier Group',
                onchange: function() {
                    filters.supplier_group = this.value || null;
					filters.supplier = []
                }
            },
			{
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'Data',
                label: 'Section, Materials, Services',
                fieldname: 'custom_section_materials_services',
                onchange: function() {
                    filters.custom_section_materials_services = this.value || null;
					filters.supplier = []
                }
            },
			{
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'MultiSelectList',
                label: 'Project',
                fieldname: 'project',
                options: 'Project',
                get_data: function (txt) {
                    return frappe.db.get_link_options("Project", txt);
                },
                onchange: function() {
                    filters.project = this.values || null;
					filters.supplier = []
                }
            },
			{
                fieldtype: 'Section Break',
            },
            {
                fieldtype: 'MultiSelectList',
                label: 'Item',
                fieldname: 'item',
				options: 'Item',
                get_data: function (txt) {
                    return frappe.db.get_link_options("Item", txt);
                },
                onchange: function() {
                    filters.item = this.values || null;
					filters.supplier = []
                }
            },
			{
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'Link',
                label: 'Item Group',
                fieldname: 'item_group',
				options: 'Item Group',
                onchange: function() {
                    filters.item_group = this.value || null;
					filters.supplier = []
                }
            },
            {
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'Select',
                label: 'Date Range',
                fieldname: 'date_range',
                default: 'Last Week',
                options: ['Last Week', 'Last Month', 'Last 3 Months', 'Last Year', 'Select Date Range'],
                onchange: function() {
                    filters.date_range = this.value || null;
					filters.supplier = []
                }
            },
            {
                fieldtype: 'Column Break',
            },
            {
                label: 'Select Date Range',
                fieldtype: 'Date Range',
                fieldname: 'selected_date_range',
                depends_on: "eval:doc.date_range == 'Select Date Range'",
                default: [frappe.datetime.month_start(), frappe.datetime.now_date()],
                onchange: function() {
                    filters.selected_date_range = this.value || null;
					filters.supplier = []
                }
            },
            {
                fieldtype: 'Section Break',
            },
            {
                fieldtype: 'Check',
                label: 'Only Suppliers with Purchase Orders',
                fieldname: 'has_po_only',
                default: 0,
                onchange: function() {
                    filters.has_po_only = this.value ? 1 : 0;
                    filters.supplier = []
                }
            },
            {
                fieldtype: 'Column Break',
            },
            {
                fieldtype: 'Check',
                label: 'Only Suppliers with Supplier Quotation',
                fieldname: 'has_sq_only',
                default: 0,
                onchange: function() {
                    filters.has_sq_only = this.value ? 1 : 0;
                    filters.supplier = []
                }
            }
        ],
        body: this.page.body,
    });
    this.form.make();

	setTimeout(() => {
		const supplierField = this.form.get_field('supplier');
		if (supplierField && supplierField.df.fieldtype === 'MultiSelectList') {
			supplierField.set_value([supplierParam]);
			supplierField.refresh();
		}
	}, 500); 

    let table_container = $("<div class='supplier-insights-table mt-3'></div>").appendTo(page.main);

    let data_table = new frappe.DataTable(table_container[0], {
        columns: [
            { name: "Supplier", width: 225, fieldtype: "Data", editable: false },
            { name: "Purchase Order", width: 200, fieldtype: "Data", editable: false },
            // { name: "Project", width: 300, fieldtype: "Data", editable: false },
        ],
        data: [],
        cellHeight: 200,
        inlineFilters: false,
        noDataMessage: "No records found",
    });

    function fetch_supplier_estimate(filters) {
        frappe.call({
            method: 'insights.insights.page.suppliers_estimate.suppliers_estimate.get_supplier_estimate',
            args: { filters },
            freeze: true,
            freeze_message: 'Fetching supplier details...',
            callback: function(r) {
                if (r.message) {
                    update_table(r.message);
                    frappe.dom.unfreeze();
                }
            }
        });
    }

    function update_table(data) {
        let table_html = `
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>Supplier</th>
                        <th>Supplier Quotation</th>
                        <th>Purchase Order</th>
                    </tr>
                </thead>
                <tbody>
                    ${data.map(row => `
                        <tr>
                            <td>
                                <strong>Supplier Code:</strong> ${row.supplier_code || ''}<br>
                                <strong>Supplier Name:</strong> ${row.supplier_name || ''}<br>
                                <strong>Section, Materials, Services:</strong> ${row.section_materials_services || ''}<br>
                                <strong>Address:</strong> ${row.address || 'NA'}<br>
                                <strong>Contact:</strong> ${row.contact || 'NA'}
                            </td>
                            <td>
                                ${row.supplier_quotation
                                    ? `Total Records: ${row.supplier_quotation.total_records || 0}<br>
                                    Total Quoted Qty: ${row.supplier_quotation.total_qty || 0} Qty<br>
                                    <button type="button" class="btn btn-secondary details-btn" 
                                        data-supplier="${row.supplier_code}" 
                                        data-supplier_name="${row.supplier_name}"
                                        data-doctype="Supplier Quotation"
                                        data-items='${JSON.stringify(row.supplier_quotation_items)}'
                                        data-details='${JSON.stringify({ 
                                            total_records: row.supplier_quotation.total_records, 
                                            total_qty: row.supplier_quotation.total_qty 
                                        })}'>
                                        Details
                                    </button>`
                                    : ``
                                }
                            </td>
                            <td>
                                ${row.purchase_order
                                    ? `Total Records: ${row.purchase_order.total_records || 0}<br>
                                    Last PO date: ${row.purchase_order.last_po_date || 0}<br>
                                    Total Order Qty: ${row.purchase_order.total_qty || 0} Qty<br>
                                    <button type="button" class="btn btn-secondary details-btn" 
                                        data-supplier="${row.supplier_code}" 
                                        data-supplier_name="${row.supplier_name}"
                                        data-doctype="Purchase Order"
                                        data-items='${JSON.stringify(row.items)}'
                                        data-details='${JSON.stringify({ 
                                            total_records: row.purchase_order.total_records, 
                                            total_qty: row.purchase_order.total_qty 
                                        })}'>
                                        Details
                                    </button>`
                                    : ``
                                    // row.projects 
                                    //     ? `Total Projects: ${row.projects.total_records || 0}<br>
                                    //     All Projects: ${row.projects.all_projects || 0}<br>
                                    //     <button class="btn btn-secondary details-btn" 
                                    //         data-supplier="${row.supplier_code}" 
                                    // 		data-supplier_name="${row.supplier_name}"
                                    //         data-doctype="Project" 
                                    //         data-details='${JSON.stringify({ 
                                    //             total_records: row.projects.total_records, 
                                    //             all_projects: row.projects.all_projects
                                    //         })}'>
                                    //         Details
                                    //     </button>`
                                    //     : ``
                                    }
                            </td>
                        </tr>
                    `).join('')}
                </tbody>
            </table>
        `;
        $(".supplier-insights-table").html(table_html);
    }

    function open_purchase_details(supplier_code, supplier_name, doctype_name, details, items) {
        let items_array = [];

        try {
            let parsed = JSON.parse(items);
            items_array = Array.isArray(parsed) ? parsed : JSON.parse(parsed);
        } catch (e) {
            console.error("Failed to parse items:", items, e);
        }
        let header_html = `
            <h5>Supplier Name - ${supplier_name}</h5>
            <h5>Supplier ID - ${supplier_code}</h5>
            <h5>Total ${doctype_name}: ${details?.total_records || 0}</h5>
            <h5>Total Order Qty: ${details?.total_qty || 0}</h5>
            <br>
            <h4>Item Details</h4>
        `;
        let items_html = "<table class='table table-sm table-bordered'><thead><tr><th>Item Code</th><th>Item Name</th><th>Qty</th></tr></thead><tbody>";

        items_array.forEach(row => {
            items_html += `<tr>
                <td>${row[1] || ''}</td>
                <td>${row[3] || ''}</td>
                <td>${row[2] || ''}</td>
            </tr>`;
        });

        items_html += "</tbody></table>";

        let d = new frappe.ui.Dialog({
            title: `${doctype_name} Details`,
            size: "large",
            fields: [{ fieldtype: "HTML", options: header_html + items_html }],
            primary_action_label: "Close",
            primary_action() { d.hide(); }
        });

        d.show();
        d.$wrapper.find('.modal-dialog').css("max-width", "1000px");
    }

    fetch_supplier_estimate(filters);
};

frappe.pages['suppliers-estimate'].on_page_show = function(wrapper) {
    if (frappe.route_options && frappe.route_options.supplier) {
        let supplierParam = frappe.route_options.supplier[0];
        frappe.route_options = null;
		filters = {};
        const supplierField = wrapper.form.get_field('supplier');
        if (supplierField && supplierField.df.fieldtype === 'MultiSelectList') {
            supplierField.set_value([supplierParam]);
            supplierField.refresh();
        }

        filters.supplier = [supplierParam];
        fetch_supplier_estimate(filters);
    }
};

File: insights/insights/page/suppliers_estimate/suppliers_estimate.py
import frappe
import json
from frappe.utils import nowdate, add_days
from datetime import datetime

@frappe.whitelist()
def get_supplier_estimate(filters):
    filters = json.loads(filters)

    date_ranges = {
        "Last Week": -7,
        "Last Month": -30,
        "Last 3 Months": -90,
        "Last Year": -365
    }

    if filters.get("date_range") == 'Select Date Range':
        start_date, end_date = filters.get("selected_date_range")
    else:
        start_date = add_days(nowdate(), date_ranges.get(filters.get("date_range"), -30))
        end_date = nowdate()
    supplier_filters = (
        {"custom_section_materials_services": ["like", "%" + filters.get("custom_section_materials_services") + "%"]}
        if filters.get("custom_section_materials_services") else {}
    )
    if filters.get("supplier_group"):
        supplier_filters["supplier_group"] = filters.get("supplier_group")
    if filters.get("supplier"):
        supplier_filters["name"] = ["in", filters.get("supplier")]
    items_list = []
    if filters.get("item"):
        items_list = filters.get("item")
    if filters.get("item_group"):
        items = frappe.get_all(
            "Item",
            filters={"item_group": filters.get("item_group")},
            pluck="name"
        )
        if items_list:
            items_list = list(set(items) & set(items_list))
        else:
            items_list.extend(items)
    suppliers = frappe.db.get_values("Supplier", supplier_filters, ["name", "supplier_name", "supplier_primary_contact", "primary_address", "custom_section_materials_services"])
    results = []
    project_dict = []
    for supplier in suppliers:
        supplier_name = supplier[1]
        supplier_quotation = list(frappe.db.get_values("Supplier Quotation", {"supplier_name" : supplier[1], "transaction_date" : ["between", [start_date, end_date]]}, ["name", "total_qty" ]))
        purchase_orders = list(frappe.db.get_values("Purchase Order", {"supplier_name" : supplier[1], "transaction_date" : ["between", [start_date, end_date]]}, ["name", "total_qty", "schedule_date"]))
        filtered_orders = []
        filtered_quotations_sq = []
        item_results = []
        sq_item_results = []
        total_qty = 0
        supplier_quotation_qty = 0
        for j in supplier_quotation:
            items = frappe.db.get_values("Supplier Quotation Item", {"parent": j[0]}, ["project", "item_code", "qty", "item_name"])
            for p in items:
                supplier_quotation_qty = supplier_quotation_qty + p[2]
                sq_item_results.append(list(p))
                if j not in filtered_quotations_sq:
                    filtered_quotations_sq.append(j)
        supplier_quotation = filtered_quotations_sq
        if len(supplier_quotation) > 0:
            sq_result = {
                "total_records": len(supplier_quotation),
                "total_qty": supplier_quotation_qty
            }
        else:
            sq_result = None
        for j in purchase_orders:
            items = frappe.db.get_values("Purchase Order Item", {"parent": j[0]}, ["project", "item_code", "qty", "item_name"])
            for p in items:
                if filters.get("project") and p[0] not in filters.get("project"):
                    continue
                if filters.get("item") and p[1] not in items_list:
                    continue
                if filters.get("item_group") and p[1] not in items_list:
                    continue
                total_qty = total_qty + p[2]
                item_results.append(list(p))
                if j not in filtered_orders:
                    filtered_orders.append(j)
        purchase_orders = filtered_orders
        if len(purchase_orders) > 0:
            total_records = len(purchase_orders)
            last_po_date = datetime.strptime(start_date, "%Y-%m-%d").date()
            for i in purchase_orders:
                items = frappe.db.get_values("Purchase Order Item", {"parent": i[0]}, ["project", "item_code"])
                for p in items:
                    if p[0] not in project_dict:
                        project_dict.append(p[0])
                if last_po_date < i[2]:
                    last_po_date = i[2]
            project_result = {
                "total_records": len(project_dict),
                "all_projects": project_dict
            }
            result = {
                "total_records": total_records,
                "total_qty": total_qty,
                "last_po_date": last_po_date,
            }
        else:
            result = None
            project_result = None
        contact_name = frappe.db.get_value("Contact", {"company_name": supplier[1]}, "name")
        contact_string = ""
        if contact_name:
            doc = frappe.get_doc("Contact", contact_name)
            if doc.email_ids:
                contact_string += ", ".join([i.email_id for i in doc.email_ids]) + "<br>"
            if doc.phone_nos:
                contact_string += ", ".join([i.phone for i in doc.phone_nos])
        if filters.get("has_sq_only") == 1 and filters.get("has_po_only") == 0 and len(supplier_quotation) > 0:
            results.append({
                "supplier_code": supplier[0],
                "supplier_name": supplier[1],
                "contact": contact_string,
                "address": supplier[3],
                "section_materials_services": supplier[4],
                "supplier_quotation":sq_result,
                "supplier_quotation_items": frappe.as_json(sq_item_results),
                "purchase_order": result,
                "projects": project_result,
                "items": frappe.as_json(item_results)
            })
        if filters.get("has_po_only") == 1 and filters.get("has_sq_only") == 0 and result:
            results.append({
                "supplier_code": supplier[0],
                "supplier_name": supplier[1],
                "contact": contact_string,
                "address": supplier[3],
                "section_materials_services": supplier[4],
                "supplier_quotation":sq_result,
                "supplier_quotation_items": frappe.as_json(sq_item_results),
                "purchase_order": result,
                "projects": project_result,
                "items": frappe.as_json(item_results)
            })
        if filters.get("has_po_only") == 0 and filters.get("has_sq_only") == 0:
            results.append({
                "supplier_code": supplier[0],
                "supplier_name": supplier[1],
                "contact": contact_string,
                "address": supplier[3],
                "section_materials_services": supplier[4],
                "supplier_quotation":sq_result,
                "supplier_quotation_items": frappe.as_json(sq_item_results),
                "purchase_order": result,
                "projects": project_result,
                "items": frappe.as_json(item_results)
            })
        if filters.get("has_po_only") == 1 and filters.get("has_sq_only") == 1 and len(supplier_quotation) > 0 and result:
            results.append({
                "supplier_code": supplier[0],
                "supplier_name": supplier[1],
                "contact": contact_string,
                "address": supplier[3],
                "section_materials_services": supplier[4],
                "supplier_quotation":sq_result,
                "supplier_quotation_items": frappe.as_json(sq_item_results),
                "purchase_order": result,
                "projects": project_result,
                "items": frappe.as_json(item_results)
            })
        result = {}
        project_result = {}
        project_dict = []
    return results


@frappe.whitelist()
def get_supplier_details(supplier_code, supplier_name, doctype, details, filters=None):
    # Fetching Doctype Details Based On Supplier And Doctype Wise
    filters = json.loads(filters)
    details = json.loads(details)

    # Date Range Logic
    date_ranges = {
        "Last Week": -7,
        "Last Month": -30,
        "Last 3 Months": -90,
        "Last Year": -365
    }
    
    if filters.get("date_range") == 'Select Date Range':
        start_date, end_date = filters.get("selected_date_range")
    else:
        start_date = add_days(nowdate(), date_ranges.get(filters.get("date_range"), -30))  # Default to Last Month if invalid
        end_date = nowdate()


    if doctype == "Purchase Order":
        return get_purchase_order_data(supplier_code, supplier_name, start_date, end_date, details)
    elif doctype == "Project":
        return get_project_data(supplier_code, supplier_name, start_date, end_date, details)


def get_purchase_order_data(supplier_code, supplier_name, start_date, end_date, details):
    # Fetching Purchase Order Details And Showcased In Table
    result = frappe.db.sql("""
        SELECT 
            poi.item_code,
            poi.item_name,
            SUM(poi.qty) AS total_qty, 
            SUM(poi.qty - poi.received_qty) AS pending_qty, 
            AVG(poi.rate) AS avg_rate, 
            SUM(poi.amount) AS total_amount
        FROM `tabPurchase Order Item` poi
        JOIN `tabPurchase Order` po ON poi.parent = po.name
        WHERE po.supplier = %s 
        AND po.transaction_date >= %s
        AND po.transaction_date <= %s
        GROUP BY poi.item_code
    """, (supplier_code, start_date, end_date), as_dict=True)

    html = f"""
        <h5>Supplier Name - {supplier_name}</h5>
        <h5>Supplier ID - {supplier_code}</h5>
        <h5>Total Purchase Order: {details.get('total_records')}</h5>
        <br>
        <h4>Item Details</h4>
        <table class="table table-bordered">
            <thead>
                <tr>
                    <th>Item Code</th>
                    <th>Item Name</th>
                    <th width="10%">Qty</th>
                    <th width="10%">Pending Qty</th>
                </tr>
            </thead>
            <tbody>
    """

    pending_qty = 0
    total_qty = 0
    for item in result:
        html += f"""
            <tr>
                <td>{item.item_code}</td>
                <td>{item.item_name}</td>
                <td>{frappe.format_value(item.total_qty, "Float")}</td>
                <td>{frappe.format_value(item.pending_qty, "Float")}</td>
            </tr>
        """
        total_qty += item.total_qty
        pending_qty += item.pending_qty

    html += f"""
        <tr style="font-weight: bold;">
            <td>Total</td>
            <td></td>
            <td>{frappe.format_value(total_qty, "Float")}</td>
            <td>{frappe.format_value(pending_qty, "Float")}</td>
    """

    html += "</tbody></table>"

    return html

def get_project_data(supplier_code, supplier_name, start_date, end_date, details):
    # Fetching Purchase Invoice Details And Showcased In Table
    result = frappe.db.sql("""
            SELECT 
                pii.item_code,
                pii.item_name,
                SUM(pii.qty) AS total_qty,
                SUM(pii.rate) AS avg_rate,
                SUM(pii.amount) AS total_amount
            FROM `tabPurchase Invoice Item` pii
            JOIN `tabPurchase Invoice` pi ON pii.parent = pi.name
            WHERE pi.supplier=%s 
            AND pi.docstatus=1 
            AND pi.posting_date >= %s
            AND pi.posting_date <= %s
            GROUP BY pii.item_code
        """, (supplier_code, start_date, end_date), as_dict=True)

    html = f"""
        <h5>Supplier Name - {supplier_name}</h5>
        <h5>Supplier ID - {supplier_code}</h5>
        <h5>Total Projects: {details.get('total_records')}</h5>
        <br>
        <h4>Item Details</h4>
        <table class="table table-bordered">
            <thead>
                <tr>
                    <th>Item Code</th>
                    <th>Item Name</th>
                    <th width="10%">Qty</th>
                    <th width="15%">Rate</th>
                    <th width="15%">Total Taxable Amount</th>
                </tr>
            </thead>
            <tbody>
    """

    total_amount = 0
    for item in result:
        html += f"""
            <tr>
                <td>{item.item_code}</td>
                <td>{item.item_name}</td>
                <td>{frappe.format_value(item.total_qty, "Float")}</td>
                <td>{frappe.format_value(item.avg_rate, "Currency")}</td>
                <td>{frappe.format_value(item.total_amount, "Currency")}</td>
            </tr>
        """
        total_amount += item.total_amount

    html += f"""
        <tr style="font-weight: bold;">
            <td>Total</td>
            <td></td>
            <td>{frappe.format_value(details.get('total_qty'), "Float")}</td>
            <td></td>
            <td>{frappe.format_value(total_amount, "Currency")}</td>
        </tr>
    """

    html += "</tbody></table>"

    return html


def get_payment_request_data(supplier_code, supplier_name, start_date, end_date, details):
    # Fetching Payment Request Details And Showcased In Table
    result = frappe.db.sql("""
        SELECT 
            pr.payment_request_type,
            pr.transaction_date, 
            pr.reference_doctype, 
            pr.reference_name,
            pr.grand_total
        FROM `tabPayment Request` pr
        WHERE
            pr.party_type='Supplier'
            AND pr.party=%s 
            AND pr.docstatus=1 
            AND pr.transaction_date >= %s 
            AND pr.transaction_date <= %s
    """, (supplier_code, start_date, end_date), as_dict=True)

    html = f"""
        <h5>Supplier Name - {supplier_name}</h5>
        <h5>Supplier ID - {supplier_code}</h5>
        <h5>Total Payment Request: {details.get('total_records')}</h5>
        <br>
        <h4>Item Details</h4>
        <table class="table table-bordered">
            <thead>
                <tr>
                    <th>Payment Request Type</th>
                    <th>Transaction Date</th>
                    <th width="25%">Reference Doctype</th>
                    <th width="25%">Reference Name</th>
                    <th width="15%">Total Amount</th>
                </tr>
            </thead>
            <tbody>
    """

    total_amount = 0
    for item in result:
        html += f"""
            <tr>
                <td>{item.payment_request_type}</td>
                <td>{frappe.format_value(item.transaction_date, "Date")}</td>
                <td>{item.reference_doctype}</td>
                <td>{item.reference_name}</td>
                <td>{frappe.format_value(item.grand_total, "Currency")}</td>
            </tr>
        """
        total_amount += item.grand_total

    html += f"""
        <tr style="font-weight: bold;">
            <td>Total</td>
            <td></td>
            <td></td>
            <td></td>
            <td>{frappe.format_value(total_amount, "Currency")}</td>
        </tr>
    """

    html += "</tbody></table>"

    return html


def get_payment_entry_data(supplier_code, supplier_name, start_date, end_date, details):
    # Fetching Payment Entry Details And Showcased In Table
    result = frappe.db.sql("""
        SELECT 
            pe.payment_type,
            pe.posting_date,
            pe.mode_of_payment,
            pe.unallocated_amount,
            pe.paid_amount
        FROM `tabPayment Entry` pe
        WHERE
            pe.party_type='Supplier'
            AND pe.party=%s 
            AND pe.docstatus=1 
            AND pe.posting_date >= %s 
            AND pe.posting_date <= %s
    """, (supplier_code, start_date, end_date), as_dict=True)

    html = f"""
        <h5>Supplier Name - {supplier_name}</h5>
        <h5>Supplier ID - {supplier_code}</h5>
        <h5>Total Payment Entry: {details.get('total_records')}</h5>
        <br>
        <h4>Item Details</h4>
        <table class="table table-bordered">
            <thead>
                <tr>
                    <th>Payment Type</th>
                    <th>Transaction Date</th>
                    <th width="20%">Mode of Payment</th>
                    <th width="20%">Unallocated Amount</th>
                    <th width="20%">Paid Amount</th>
                </tr>
            </thead>
            <tbody>
    """

    unllocated_amount = 0
    total_amount = 0
    for item in result:
        html += f"""
            <tr>
                <td>{item.payment_type}</td>
                <td>{frappe.format_value(item.posting_date, "Date")}</td>
                <td>{item.mode_of_payment}</td>
                <td>{frappe.format_value(item.unallocated_amount, "Currency")}</td>
                <td>{frappe.format_value(item.paid_amount, "Currency")}</td>
            </tr>
        """
        unllocated_amount += item.unallocated_amount
        total_amount += item.paid_amount

    html += f"""
        <tr style="font-weight: bold;">
            <td>Total</td>
            <td></td>
            <td></td>
            <td>{frappe.format_value(unllocated_amount, "Currency")}</td>
            <td>{frappe.format_value(total_amount, "Currency")}</td>
        </tr>
    """

    html += "</tbody></table>"

    return html